lib/ecto/repo/schema.ex

```diff
@@ -199,8 +199,18 @@ defmodule Ecto.Repo.Schema do
           autogenerate_id(autogen_id, changes, return_types, return_sources, adapter)
         {changes, autogen} =
           dump_changes!(:insert, Map.take(changes, fields), schema, extra, dumper, adapter)
+
+        {changeset, changed_fields} =
+          case on_conflict do
+            :replace_all ->
+              lift_values_from_changeset(changeset, struct, fields)
+
+           _ ->
+            {changeset, Keyword.keys(changes)}
```

Comment:
This is not entirely correct - similar logic should be applied to `replace_app_except_primary_keys` option.

I wonder if maybe this logic should be moved into `on_conflict`, but then this would cause problems in sharing it between `insert` and `insert_all`.